temporalio/converter.py

```diff
@@ -594,6 +596,72 @@ def default() -> DataConverter:
     return _default
 
 
+def encode_search_attributes(
+    attrs: temporalio.common.SearchAttributes,
+    api: temporalio.api.common.v1.SearchAttributes,
+) -> None:
+    """Convert search attributes into an API message.
+
+    Args:
+        attrs: Search attributes to convert.
+        api: API message to set converted attributes on.
+    """
+    for k, v in attrs.items():
+        api.indexed_fields[k].CopyFrom(encode_search_attribute_values(v))
+
+
+def encode_search_attribute_values(
+    vals: List[temporalio.common.SearchAttributeValue],
+) -> temporalio.api.common.v1.Payload:
+    """Convert search attribute values into a payload.
+
+    Args:
+        vals: List of values to convert.
+    """
+    if not isinstance(vals, list):
+        raise TypeError("Search attribute values must be lists")
+    # Convert dates to strings
+    safe_vals = []
+    for v in vals:
```

Comment:
You should assert that vals is a homogenous list

Reply:
Could also limit this at the type level.
We should have done the same in TS @lorensr (still not too late).

```ts
type SearchAttributeValueArray = string[] | number[] | boolean[] | Date[];
type SearchAttributes = Record<string, SearchAttributeVAlueArray>;
```

Reply:
Sounds good, will do.

Reply:
Makes sense, will do

Reply:
I am afraid that `Union[List[str], List[int], List[float], List[bool], List[datetime]]` doesn't work as we'd like. You can't assign an empty list to it for example (this is a known MyPy and Python typing issue, see issue 3283 in https://github.com/python/mypy). We're gonna have to keep the types as is and only do runtime checks.